src/Vinyl-s3.js

```diff
@@ -83,84 +105,125 @@ class VinylS3 {
       // Pop the oldest key off the top of the array FIFO style
       const key = self.keys.pop();
 
-      // Start creating our Vinyl object params. Remove any prefix path like stage or prod
-      const regex = new RegExp(`^${self.relativePath}`);
-      const vinylParams = { path: key.replace(regex, "") };
-
-      // Fetch the data from S3
-      try {
-        const s3Object = await self.s3
-          .getObject({
-            Bucket: self.options.Bucket,
-            Key: key,
-          })
-          .promise();
-
-        // Set the directory or file details, as determined by the key, in our Vinyl object params
-        if (key.match(/\/$/) > "") {
-          // Is directory so set mode and leave contents as null
-          vinylParams.stat = { mode: 16384 };
-        } else {
-          // Is file, so set contents
-          vinylParams.contents = Buffer.from(s3Object.Body);
+      // Start creating our Vinyl object params. Remove any prefix path like /src/stage or /dist/prod
+      const prefix = new RegExp(`^${self.options.base}/`);
+      const vinylParams = { path: key.replace(prefix, "") };
+
+      // TODO: This is absolutely BRUTE-FORCE, copy/paste duplicated, and FUGLY code that must be refactored!
+      if (head) {
+        try {
+          const s3Object = await self.s3
+            .headObject({
+              Bucket: self.options.bucket,
+              Key: key,
+            })
+            .promise();
+
+          // Set the directory or file details, as determined by the key, in our Vinyl object params
+          if (key.match(/\/$/) > "") {
+            // Is directory so set mode and leave contents as null
+            vinylParams.stat = { mode: 16384 };
+          } else {
+            // Is file, so set contents
+            vinylParams.stat = { mode: 0 };
+            // ! JSON.parse() required! S3 returns in the format ""1f077ad7e6b4bf07910a0c5b9f654e6c"" but we do not know why yet
+            vinylParams.eTag = JSON.parse(s3Object.ETag);
+            vinylParams.size = s3Object.ContentLength;
+          }
+          // Create a new Vinyl object
+          const vinyl = new Vinyl(vinylParams);
+
+          // Push the Vinyl object into the stream
+          readable.push(vinyl);
+        } catch (err) {
+          console.error("Vinyl-s3._read()", err, key, vinylParams);
+          throw err;
+        }
+      } else {
+        // Fetch the data from S3
+        try {
+          const s3Object = await self.s3
+            .getObject({
+              Bucket: self.options.bucket,
+              Key: key,
+            })
+            .promise();
+
+          // Set the directory or file details, as determined by the key, in our Vinyl object params
+          if (key.match(/\/$/) > "") {
+            // Is directory so set mode and leave contents as null
+            vinylParams.stat = { mode: 16384 };
```

Comment:
Perhaps these stat mode values could become an Enum with more legible names?

Reply:
Yes. Zero is the wrong one for file. My Google Fu was way off that day and I was struggling to find documentation of correct values. I will add a backlog ticket.

https://dev.azure.com/techsmarts/Web%20Producer/_workitems/edit/717